File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.js
function solution(maps) {
    const height = maps.length;
    const width = maps[0].length;
    
    // 1. 방향 벡터 (상, 하, 좌, 우)
    const dx = [0, 0, 1, -1];
    const dy = [1, -1, 0, 0];
    

    let visited = Array.from({ length: height }, () => Array(width).fill(0));
    let queue = [];
    queue.push([0, 0]); 
    visited[0][0] = 1;  
    
    while (queue.length > 0) {
        let [y, x] = queue.shift(); 
        
        for (let i = 0; i < 4; i++) {
            let ny = y + dy[i];
            let nx = x + dx[i];
            
            if (ny >= 0 && ny < height && nx >= 0 && nx < width) {
                if (maps[ny][nx] === 1 && visited[ny][nx] === 0) {
                    visited[ny][nx] = 1;
                    maps[ny][nx] = maps[y][x] + 1;
                    queue.push([ny, nx]);
                }
            }
        }
    }

    const answer = maps[height - 1][width - 1];
    
    return answer === 1 ? -1 : answer;
}

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
from collections import deque

def solution(maps):
    n = len(maps)
    m = len(maps[0])
    visited = [[False for _ in range(m)] for _ in range(n)]
    
    # 처음 방문
    q = deque()
    q.append((0, 0))
    visited[0][0] = True
    
    # 이동방향
    dx = [1, -1, 0, 0]
    dy = [0, 0, 1, -1]
    
    while q:
        y, x = q.popleft()
        for i in range(4):
            nx = x + dx[i]
            ny = y + dy[i]
            # 범위에 벗어나지 않고 1이면
            if 0 <= nx < m and 0 <= ny < n and maps[ny][nx] == 1:
                if not visited[ny][nx]:
                    # 방문처리
                    visited[ny][nx] = True
                    q.append((ny, nx))
                    maps[ny][nx] = maps[y][x] + 1

    if maps[n-1][m-1] == 1:
        return -1
    
    return maps[n-1][m-1]
